### PR TITLE
feat: support BLS12-381 (based on `bls12_381` crate)

### PR DESCRIPTION
**Pull request by nulltea**
Add curve BLS12-381. Targets `ff = "0.13.0"`.

### Sources:
- https://github.com/privacy-scaling-explorations/bls12_381 (all types, majority of trait impls, and tests)
- https://github.com/privacy-scaling-explorations/halo2curves/blob/support_bls12-381/src/bls12_381/mod.rs (endomorphism)

The decision to copy implementation from `bls12_381` crate is explained by the numerous additional trait impls needed to be compatible with `halo2-lib`. See bellow.

### Additional changes
- implement `ff::PrimeField` for [Fp](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/fp.rs#L366-L396C2), [Scalar](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/scalar.rs#L723-L752), [Fp2](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/fp2.rs#L625-L680).
- implement misc traits for halo2lib compatibility for [Fp](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/fp.rs#L960-L1000), [Scalar](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/scalar.rs#L825-L856), [Fp2](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/fp2.rs#L682-L696)
- implement ff::Field for [Fp2](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/fp2.rs#L416-L599C6) including [`sqrt_ratio`](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/fp2.rs#L495-L599C6).
- implement `CurveAffine` and `CurveAffineExt` for [G1](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/g1.rs#L491-L539) and [G2](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/g2.rs#L1341-L1388)
- switch to LE encoding for [Fp](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/fp.rs#L413-L517), [G1Affine](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/g1.rs#L1198-L1231) and [G2Affine](https://github.com/timoftime/halo2curves/blob/support_bls12-381/src/bls12_381/g2.rs#L1254-L1284) for compatibility with halo2lib 
   - BE encoding is kept in a separate methods to remain compatible with test vectors, and since BE is more commonly used for BLS12 (e.g. Ethereum)
- use local `arithmetic.rs` funcs and a `derive` module instead of https://github.com/privacy-scaling-explorations/bls12_381/blob/main/src/util.rs

### CODE REVIEW

**Comment by jonathanpwang** [comment]
Thanks for this! 
Will review it soon (and then your `halo2-lib` PRs).

File: src/bls12_381/fp6.rs
Comment (jonathanpwang) [inline src/bls12_381/fp6.rs]: RHS should probably be $\mathbb{F}_{p^2}[v] / (v^3 - u - 1)$

File: src/bls12_381/fp12.rs
Comment (jonathanpwang) [inline src/bls12_381/fp12.rs]: $\mathbb{F}_{p^6}[w] / (w^2 - v)$

File: src/bls12_381/fp2.rs
Comment (jonathanpwang) [inline src/bls12_381/fp2.rs:713]: Is this trait needed? because right now `ZETA` is not a 3rd root of unity

File: src/bls12_381/fp2.rs
Comment (jonathanpwang) [inline src/bls12_381/fp2.rs:624]: Is this needed somewhere? It is weird to implement `PrimeField` for `Fp2` because it is not a prime field. 

I don't think this should be needed for `halo2-ecc` usage.

File: src/bls12_381/fp6.rs
Comment (nulltea) [reply src/bls12_381/fp6.rs]: this comes from zkcrypto crate: https://github.com/zkcrypto/bls12_381/blob/main/src/fp6.rs#L11, but I see that the definition from BN  related types is how you suggest, so I'll update documentation for BLS12 types accordingly

File: src/bls12_381/fp2.rs
Comment (nulltea) [reply src/bls12_381/fp2.rs:713]: `CurveExt` trait requires it https://github.com/zcash/pasta_curves/blob/main/src/arithmetic/curves.rs#L32 since `Fp2` is used as base field for `G2Affine`

File: src/bls12_381/fp2.rs
Comment (nulltea) [reply src/bls12_381/fp2.rs:624]: Referring to the previous comment, since `Fp2` must implement `WithSmallOrderMulGroup` to be used as `G2Affine` base it's also must implement `PrimeField` https://github.com/zkcrypto/ff/blob/main/src/lib.rs#L358

Comment (nulltea) [reply src/bls12_381/fp2.rs:624]: I agree that it's might become confusion, so perhaps I can add comments explaining the technical reasons why `PrimeField` and  `WithSmallOrderMulGroup` traits needed to be implemented for Fp2

Comment (jonathanpwang) [reply src/bls12_381/fp2.rs:624]: OK, I see it is because `CurveExt` extends `PrimeCurveAffine`, which is unfortunate and should ideally not be there. I see that `bn256::Fq2` is also following the same pattern, so it's a common issue. 

Please just add a comment explaining then.

File: src/bls12_381/fp2.rs
Comment (jonathanpwang) [inline src/bls12_381/fp2.rs:650]: since you already computed `ROOT_OF_UNITY`, can you also compute `ROOT_OF_UNITY^{-1}` and fill it in here? And add a unit test that they multiply to 1.

File: src/bls12_381/fp2.rs
Comment (jonathanpwang) [inline src/bls12_381/fp2.rs:672]: Please calculate `2^{-1}` and fill in here.

File: src/bls12_381/fp2.rs
Comment (jonathanpwang) [inline src/bls12_381/fp2.rs]: not sure if you use this, but if we follow bn256::Fq2, then we can put `c0: Fp::ZETA ^ 2, c1: Fp::ZERO` here: https://github.com/axiom-crypto/halo2curves/blob/55cab0cb9c2439758d09743cbd6e5946cf10876f/src/bn256/fq2.rs#L530

(I am not sure why they use ZETA^2, but both are 3rd roots of unity.)

This would make the trait implementation accurate mathematically.

**Review by jonathanpwang — CHANGES_REQUESTED** [review]
Due to the necessity of some prime field traits on `fp2`, request to at least put in the correct constants when possible. After that good to merge.

**Review by jonathanpwang — APPROVED** [review]
LGTM after updating sha2 and digest import versions

**Pull request merged**
